cpp/include/raft/spectral/lanczos.hpp

```diff
@@ -0,0 +1,1228 @@
+ *
+ *  @param n Matrix dimension.
+ *  @param iter Current Lanczos iteration.
+ *  @param iter_new Lanczos iteration after restart.
+ *  @param shiftUpper Pointer to upper bound for unwanted
+ *    region. Value is ignored if less than *shiftLower. If a
+ *    stronger upper bound has been found, the value is updated on
+ *    exit.
+ *  @param shiftLower Pointer to lower bound for unwanted
+ *    region. Value is ignored if greater than *shiftUpper. If a
+ *    stronger lower bound has been found, the value is updated on
+ *    exit.
+ *  @param alpha_host (Input/output, host memory, iter entries)
+ *    Diagonal entries of Lanczos system.
+ *  @param beta_host (Input/output, host memory, iter entries)
+ *    Off-diagonal entries of Lanczos system.
+ *  @param V_host (Output, host memory, iter*iter entries)
+ *    Orthonormal transform used to obtain restarted system. Matrix
+ *    dimensions are iter x iter.
+ *  @param work_host (Output, host memory, 4*iter entries)
+ *    Workspace.
+ *  @param lanczosVecs_dev (Input/output, device memory, n*(iter+1)
+ *    entries) Lanczos vectors. Vectors are stored as columns of a
+ *    column-major matrix with dimensions n x (iter+1).
+ *  @param work_dev (Output, device memory, (n+iter)*iter entries)
+ *    Workspace.
+ */
+template <typename IndexType_, typename ValueType_>
+static int lanczosRestart(
+  handle_t const &handle, IndexType_ n, IndexType_ iter, IndexType_ iter_new,
+  ValueType_ *shiftUpper, ValueType_ *shiftLower,
+  ValueType_ *__restrict__ alpha_host, ValueType_ *__restrict__ beta_host,
+  ValueType_ *__restrict__ V_host, ValueType_ *__restrict__ work_host,
+  ValueType_ *__restrict__ lanczosVecs_dev, ValueType_ *__restrict__ work_dev,
+  bool smallest_eig) {
+  // -------------------------------------------------------
+  // Variable declaration
+  // -------------------------------------------------------
+
+  // Useful constants
+  const ValueType_ zero = 0;
+  const ValueType_ one = 1;
+
+  auto cublas_h = handle.get_cublas_handle();
+  auto stream = handle.get_stream();
+
+  // Loop index
+  IndexType_ i;
+
+  // Number of implicit restart steps
+  //   Assumed to be even since each call to Francis algorithm is
+  //   equivalent to two calls of QR algorithm
+  IndexType_ restartSteps = iter - iter_new;
+
+  // Ritz values from Lanczos method
+  ValueType_ *ritzVals_host = work_host + 3 * iter;
+  // Shifts for implicit restart
+  ValueType_ *shifts_host;
+
+  // Orthonormal matrix for similarity transform
+  ValueType_ *V_dev = work_dev + n * iter;
+
+  // -------------------------------------------------------
+  // Implementation
+  // -------------------------------------------------------
+
+  // Compute Ritz values
+  memcpy(ritzVals_host, alpha_host, iter * sizeof(ValueType_));
+  memcpy(work_host, beta_host, (iter - 1) * sizeof(ValueType_));
+  Lapack<ValueType_>::sterf(iter, ritzVals_host, work_host);
+
+  // Debug: Print largest eigenvalues
```

Comment:
I'd rather keep debug-related comments, if you don't mind. When debug time comes, they turn out to be very useful.

Reply:
Fair point. I think we should revisit once we have figured out how we should do logging in a unified way across cugraph/cuml. What do you think? For example, it would be really nice to have this always log to debug (or when debug mode is on).

Reply:
OK. We need to be careful, because many times "DEBUG"-decorated code comes with a bunch of necessary dependencies (`#include <iostream>` , etc.) which must be `#ifdef`-ed out (otherwise they pollute the namespace and increase compilation time). Plus...`#ifdef`-ed out code is ugly (uglier then...just comments). 

But I agree with adding more structure and discipline in the (near) future, when dealing with such things.